types/math/dec.go

```diff
@@ -0,0 +1,126 @@
+package math
+
+import (
+	"github.com/cockroachdb/apd/v2"
+	"github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+type Dec struct {
+	dec apd.Decimal
+}
+
+var dec128Context = apd.Context{
+	Precision:   34,
+	MaxExponent: apd.MaxExponent,
+	MinExponent: apd.MinExponent,
+	Traps:       apd.DefaultTraps,
+}
+
+func NewDecFromString(s string) (Dec, error) {
+	d, _, err := apd.NewFromString(s)
+	if err != nil {
+		return Dec{}, err
+	}
+	return Dec{*d}, nil
+}
+
+func NewDecFromInt64(x int64) Dec {
+	var res Dec
+	res.dec.SetInt64(x)
+	return res
+}
+
+func (x Dec) Add(y Dec) (Dec, error) {
```

Comment:
```suggestion
// Add returns new Dec with value `x+y` without mutating any argument and error if 
// there is an overflow.
func (x Dec) Add(y Dec) (Dec, error) {
```

Reply:
Shall we add similar comment to other functions?

Reply:
This is done!